storage/src/main/java/tech/pegasys/artemis/storage/V2MapDatabase.java

```diff
@@ -0,0 +1,291 @@
+        .forEach(
+            root -> {
+              addHotBlock(root, store.getBlock(root));
+              hotStatesByRoot.put(root, store.getBlockState(root));
+            });
+    checkpointStates.put(
+        store.getJustifiedCheckpoint(),
+        store.getBlockState(store.getJustifiedCheckpoint().getRoot()));
+  }
+
+  @Override
+  public void insert(final Transaction transaction) {
+    final Checkpoint previousFinalizedCheckpoint = finalizedCheckpoint.get();
+    final Checkpoint newFinalizedCheckpoint = transaction.getFinalizedCheckpoint();
+    time.set(transaction.getTime());
+    genesisTime.set(transaction.getGenesisTime());
+    finalizedCheckpoint.set(newFinalizedCheckpoint);
+    justifiedCheckpoint.set(transaction.getJustifiedCheckpoint());
+    bestJustifiedCheckpoint.set(transaction.getBestJustifiedCheckpoint());
+    checkpointStates.putAll(transaction.getCheckpointStates());
+    latestMessages.putAll(transaction.getLatestMessages());
+
+    transaction.getBlocks().forEach(this::addHotBlock);
+    hotStatesByRoot.putAll(transaction.getBlockStates());
+
+    if (previousFinalizedCheckpoint == null
+        || !previousFinalizedCheckpoint.equals(newFinalizedCheckpoint)) {
+      recordFinalizedBlocks(newFinalizedCheckpoint);
+      pruneCheckpointStates(newFinalizedCheckpoint);
+      pruneHotBlocks(newFinalizedCheckpoint);
+    }
+    db.commit();
+  }
+
+  private void addHotBlock(final Bytes32 root, final BeaconBlock block) {
+    hotBlocksByRoot.put(root, block);
+    addToHotRootsBySlotCache(root, block);
+  }
+
+  private void recordFinalizedBlocks(final Checkpoint newFinalizedCheckpoint) {
+    LOG.debug(
+        "Record finalized blocks for epoch {} starting at block {}",
+        newFinalizedCheckpoint.getEpoch(),
+        newFinalizedCheckpoint.getRoot());
+    final UnsignedLong highestFinalizedSlot =
+        finalizedRootsBySlot.isEmpty() ? UnsignedLong.ZERO : finalizedRootsBySlot.lastKey();
+    Bytes32 newlyFinalizedBlockRoot = newFinalizedCheckpoint.getRoot();
+    BeaconBlock newlyFinalizedBlock = hotBlocksByRoot.get(newlyFinalizedBlockRoot);
+    while (newlyFinalizedBlock != null
+        && newlyFinalizedBlock.getSlot().compareTo(highestFinalizedSlot) > 0) {
+      LOG.debug(
+          "Recording finalized block {} at slot {}",
+          newlyFinalizedBlock.getSlot(),
+          newlyFinalizedBlockRoot);
+      finalizedRootsBySlot.put(newlyFinalizedBlock.getSlot(), newlyFinalizedBlockRoot);
+      finalizedBlocksByRoot.put(newlyFinalizedBlockRoot, newlyFinalizedBlock);
+      final Optional<BeaconState> finalizedState = getState(newlyFinalizedBlockRoot);
+      if (finalizedState.isPresent()) {
+        finalizedStatesByRoot.put(newlyFinalizedBlockRoot, finalizedState.get());
+      }
+      newlyFinalizedBlockRoot = newlyFinalizedBlock.getParent_root();
+      newlyFinalizedBlock = hotBlocksByRoot.get(newlyFinalizedBlockRoot);
+    }
+
+    if (newlyFinalizedBlock == null) {
```

Comment:
Should we throw an exception here?

Reply:
ooo that opens a can of worms. If for any reason `insert` fails we won't commit the db and the change will be lost.  But the changes will still have happened in the in-memory `Store` and now we're out of sync. I've logged https://pegasys1.atlassian.net/browse/BC-188 to work this out.